Add HLS support for "stpp.*" codec support for SMPTE-TT fmp4 subtitle…

stevemayhew opened the pull request:
We will enter a bug for this shortly. 

ExoPlayer needs a codec to decide among WEBVTT and TTML decoder mimeType.
Apple describes IMSC1 in MP4 in
[RFC-8216 Section 3.6](https://tools.ietf.org/html/draft-pantos-hls-rfc8216bis-04#section-3.6).

The DASH manifest specifies the SMPTE-TT captions in the codecs in the manifest
(from W3C [TTML Profiles for Internet Media Subtitles and Captions 1.1](https://www.w3.org/TR/ttml-imsc1.1/#general-0).
DASH just doesn't require the rendition linking, but HLS does.

Apple implies the CODECS attribute of the variant needs to be do this.  That is
with SHOULD and MAY language to imply the codec to use for it in the
[Authoring Guidelines](https://developer.apple.com/documentation/http_live_streaming/hls_authoring_specification_for_apple_devices)

This change defaults to WebVTT if no codec is specifed (same as current behavior) otherwise it picks it from the variants
referencing the media.

stevemayhew committed:
Add HLS support for "stpp.*" codec support for SMPTE-TT fmp4 subtitle tracks

ExoPlayer needs a codec to decide among WEBVTT and TTML decoder mimeType.
Apple describes IMSC1 in MP4 in
[RFC-8216 Section 3.6](https://tools.ietf.org/html/draft-pantos-hls-rfc8216bis-04#section-3.6).

The DASH manifest specifies the SMPTE-TT captions in the codecs in the manifest
(from W3C [TTML Profiles for Internet Media Subtitles and Captions 1.1](https://www.w3.org/TR/ttml-imsc1.1/#general-0).
DASH just doesn't require the rendition linking, but HLS does.

Apple implies the CODECS attribute of the variant needs to be do this.  That is
with SHOULD and MAY language to imply the codec to use for it in the
[Authoring Guidelines](https://developer.apple.com/documentation/http_live_streaming/hls_authoring_specification_for_apple_devices)

This change defaults to WebVTT if no codec is specifed (same as current behavior) otherwise it picks it from the variants
referencing the media.

jparekh-tivo commented:
Linked issue is - https://github.com/google/ExoPlayer/issues/7185

stevemayhew committed:
Cleanup to use getCodecsOfType()

Use the `Util.getCodecsOfType()` method rather than hard coded codec strings

ojw28 commented:
Please could you get rid of the 7 unnecessary files being added as part of this pull request?

stevemayhew commented on library/hls/src/main/java/com/google/android/exoplayer2/source/hls/playlist/HlsPlaylistParser.java:
```diff
@@ -458,7 +458,18 @@ private static HlsMasterPlaylist parseMasterPlaylist(LineIterator iterator, Stri
           break;
         case TYPE_SUBTITLES:
-          formatBuilder.setSampleMimeType(MimeTypes.TEXT_VTT).setMetadata(metadata);
+          sampleMimeType = MimeTypes.TEXT_VTT;
```
Default is original behavior (assume VTT)...

stevemayhew committed:
Fix merge mistake, wrong track type.

Fix mistake from the backport merge, should get TEXT track type for subtitles.  Also update testcase to match pattern in `dev-v2` branch.

stevemayhew commented:
@ojw28 Sorry, there was also an error in the merge from our commit (that is release branch based) that I fixed (track type).  The test case passes and covers the default case (select VTT if no matching codecs in the variant CODEC) works fine.   We could also add the explicit CODEC of "wvtt", I did not see any reference to this though outside of Apple guidelines.
